File: ui/src/types/preferences.ts
/** Content-sync configuration (mirrors `lazybones_store::SyncConfig`). The
 *  git-backed sync repo that carries authored docs/skills/tasks/templates/
 *  workflows between machines. */
export interface SyncConfig {
  /** Master switch for the automatic behaviours (auto-pull on boot + periodic
   *  auto-push). Manual pull/push still work when this is off. */
  enabled: boolean;
  /** Git remote URL of the cloud sync repo, or `null`/empty => not configured. */
  remote: string | null;
  /** Branch to sync on; `null` is treated as `"main"`. */
  branch: string | null;
  /** Absolute path to the local checkout; `null` lets the daemon derive one. */
  dir: string | null;
  /** Push the export automatically after store changes. */
  auto_push: boolean;
  /** Pull + import automatically on daemon boot. */
  auto_pull: boolean;
}

/** The single global user-preferences record (mirrors `lazybones_store::Preferences`).
 *  Operator UI choices that follow the operator across browsers/devices. */
export interface Preferences {
  /** IANA timezone name, or `null` to follow the browser's zone. */
  timezone: string | null;
  /** UI theme: `"light" | "dark" | "system"`, or `null` for system. */
  theme: string | null;
  /** Content-sync config, or `null` until the operator sets it up. */
  sync: SyncConfig | null;
  /** RFC3339 timestamp of the last write (empty until first saved). */
  updated_at: string;
}

/** `PUT /settings/preferences` body. An omitted/empty field clears that
 *  preference (reverts to its default). */
export interface PreferencesDraft {
  timezone?: string | null;
  theme?: string | null;
  sync?: SyncConfig | null;
}

File: crates/lazybones-api/src/routes/preferences.rs
//! `GET /settings/preferences` + `PUT /settings/preferences` — the single
//! global user-preferences record (display timezone, UI theme).
//!
//! These are operator UI choices that should follow the operator across
//! browsers/devices, rather than living in each browser's `localStorage`.
//! Reads return the stored preferences (or usable defaults when unset). Writes
//! require `Author`.

use axum::Json;
use axum::extract::State;
use lazybones_auth::Capability;
use lazybones_store::Preferences;

use crate::dto::PreferencesBody;
use crate::error::ApiResult;
use crate::extract::Session;
use crate::state::AppState;

/// `GET /settings/preferences` — the current preferences, or the defaults if the
/// operator has never saved any. Open read (local single-user daemon).
pub async fn get_preferences(State(state): State<AppState>) -> ApiResult<Json<Preferences>> {
    let prefs = state.store.get_preferences().await?.unwrap_or_default();
    Ok(Json(prefs))
}

/// `PUT /settings/preferences` — replace the preferences record. Requires
/// `Author`. An omitted field clears that preference (reverts to its default).
pub async fn put_preferences(
    State(state): State<AppState>,
    session: Session,
    Json(body): Json<PreferencesBody>,
) -> ApiResult<Json<Preferences>> {
    session.require(Capability::Author, "author", "preferences")?;

    // An empty timezone string means "follow the browser" — normalise to None.
    let timezone = body
        .timezone
        .map(|tz| tz.trim().to_owned())
        .filter(|tz| !tz.is_empty());

    // Normalise the sync config: canonicalise the remote (shorthand / https / ssh
    // → a gh-auth-friendly https URL) so it's stored in the working form and the
    // UI reflects what it resolved to. An all-whitespace remote means "not
    // configured", so drop the whole config when it carries no usable URL.
    let sync = body.sync.map(normalize_sync).filter(lazybones_store::SyncConfig::is_configured);

    let prefs = Preferences {
        timezone,
        theme: body.theme,
        sync,
        updated_at: state.store.now(),
    };
    Ok(Json(state.store.put_preferences(&prefs).await?))
}

/// Canonicalise a sync config's remote URL in place (empty remote → cleared).
fn normalize_sync(mut cfg: lazybones_store::SyncConfig) -> lazybones_store::SyncConfig {
    let normalized = lazybones_engine::sync::normalize_remote(cfg.remote.as_deref().unwrap_or_default());
    cfg.remote = if normalized.is_empty() { None } else { Some(normalized) };
    cfg
}
